filestack/utils/intelligent_ingestion.py

```diff
@@ -0,0 +1,435 @@
+
+        success = True
+        try:
+            backend_resp = requests.post(
+                UPLOAD_HOST + '/multipart/upload',
+                data={
+                    'apikey': job['apikey'],
+                    'part': job['part'],
+                    'size': job['size'],
+                    'md5': b64encode(hashlib.md5(chunk).digest()).strip(),
+                    'uri': job['uri'],
+                    'region': job['region'],
+                    'upload_id': job['upload_id'],
+                    'store_location': job['store_location'],
+                    'multipart': True,
+                    'offset': job['offset']
+                },
+                files={'file': (job['filename'], '', None)},
+                headers=HEADERS
+            )
+            if not backend_resp.ok:
+                raise ResponseNotOk('Incorrect backend response %s', backend_resp)
+
+            backend_data = backend_resp.json()
+            try:
+                s3_resp = requests.put(
+                    backend_data['url'],
+                    headers=backend_data['headers'],
+                    data=chunk
+                )
+            except Exception as e:
+                log.warning('Upload to S3 failed %s', e)
+                raise S3UploadException(str(e))
+
+            if not s3_resp.ok:
+                raise ResponseNotOk('Incorrect S3 response %s', s3_resp)
+
+        except ResponseNotOk:
+            delay = delay * 1.3 or 1
+            success = False
+        except S3UploadException:
+            delay = 0
+            success = False
+        except Exception as e:
+            delay = 0
+            log.error('Request to backend failed %s', e)
+            success = False
+
+        response_q.put({
+            'worker': 'uploader',
+            'chunk': job['chunk'],
+            'part': job['part'],
+            'offset': job['offset'],
+            'size': job['size'],
+            'success': success,
+            'delay': delay
+        })
+
+        log.info(
+            'Uploader finished chunk %s for part %s. Success: %s',
+            job['chunk'], job['part'], success
+        )
+
+
+def commit_part(commit_q, response_q):
+    log.info('Committer ready')
+    while True:
+        job = commit_q.get(block=True)
+
+        if job == 'die':
+            break  # we need a way to stop it in tests (other than terminate())
+
+        log.info('Committer got job for part %s', job['part'])
+        log.debug('Job details: %s', job)
+
+        requests.post(
+            UPLOAD_HOST + '/multipart/commit',
+            data={
+                'apikey': job['apikey'],
+                'uri': job['uri'],
+                'region': job['region'],
+                'upload_id': job['upload_id'],
+                'size': job['size'],
+                'part': job['part'],
+                'store_location': job['store_location']
+            },
+            files={'file': (job['filename'], '', None)},
```

Comment:
when you specify `files` parameter, `requests` will send `multipart/form-data` POST, which is what we need here :) More info [here](http://docs.python-requests.org/en/master/user/quickstart/#post-a-multipart-encoded-file)